chainerx_cc/chainerx/cuda/numeric.cuh

```diff
@@ -22,89 +22,28 @@ __device__ inline bool IsInf(cuda::Float16 value) { return value.IsInf(); }
-
-template <typename T>
-__device__ inline T Sin(T x) {
-    return std::sin(x);
-}
-
-__device__ inline cuda::Float16 Sin(cuda::Float16 x) { return cuda::Float16{std::sin(static_cast<float>(x))}; }
-
-template <typename T>
-__device__ inline T Cos(T x) {
-    return std::cos(x);
-}
-
-__device__ inline cuda::Float16 Cos(cuda::Float16 x) { return cuda::Float16{std::cos(static_cast<float>(x))}; }
-
-template <typename T>
-__device__ inline T Tan(T x) {
-    return std::tan(x);
-}
-
-__device__ inline cuda::Float16 Tan(cuda::Float16 x) { return cuda::Float16{std::tan(static_cast<float>(x))}; }
-
-template <typename T>
-__device__ inline T Arcsin(T x) {
-    return std::asin(x);
-}
-
-__device__ inline cuda::Float16 Arcsin(cuda::Float16 x) { return cuda::Float16{std::asin(static_cast<float>(x))}; }
-
-template <typename T>
-__device__ inline T Arccos(T x) {
-    return std::acos(x);
-}
-
-__device__ inline cuda::Float16 Arccos(cuda::Float16 x) { return cuda::Float16{std::acos(static_cast<float>(x))}; }
-
-template <typename T>
-__device__ inline T Arctan(T x) {
-    return std::atan(x);
-}
-
-__device__ inline cuda::Float16 Arctan(cuda::Float16 x) { return cuda::Float16{std::atan(static_cast<float>(x))}; }
-
-template <typename T>
-__device__ inline T Exp(T x) {
-    return std::exp(x);
-}
-
-__device__ inline cuda::Float16 Exp(cuda::Float16 x) { return x.Exp(); }
-
-template <typename T>
-__device__ inline T Log(T x) {
-    return std::log(x);
-}
-
-__device__ inline cuda::Float16 Log(cuda::Float16 x) { return x.Log(); }
-
-template <typename T>
-__device__ inline T Sqrt(T x) {
-    return std::sqrt(x);
-}
-
-__device__ inline cuda::Float16 Sqrt(cuda::Float16 x) { return x.Sqrt(); }
+#define DECLARE_CUDA_FLOAT16_FALLBACK_UNARY(name, expr)         \
+    template <typename T>                                       \
+    __device__ inline T name(T x) {                             \
+        return expr;                                            \
+    }                                                           \
+    __device__ inline cuda::Float16 name(cuda::Float16 input) { \
+        float x = static_cast<float>(input);                    \
+        return cuda::Float16{expr};                             \
+    }
+
+DECLARE_CUDA_FLOAT16_FALLBACK_UNARY(Ceil, std::ceil(x))
+DECLARE_CUDA_FLOAT16_FALLBACK_UNARY(Floor, std::floor(x))
+DECLARE_CUDA_FLOAT16_FALLBACK_UNARY(Tanh, std::tanh(x))
+DECLARE_CUDA_FLOAT16_FALLBACK_UNARY(Sin, std::sin(x))
+DECLARE_CUDA_FLOAT16_FALLBACK_UNARY(Cos, std::cos(x))
+DECLARE_CUDA_FLOAT16_FALLBACK_UNARY(Tan, std::tan(x))
+DECLARE_CUDA_FLOAT16_FALLBACK_UNARY(Arcsin, std::asin(x))
+DECLARE_CUDA_FLOAT16_FALLBACK_UNARY(Arccos, std::acos(x))
+DECLARE_CUDA_FLOAT16_FALLBACK_UNARY(Arctan, std::atan(x))
+DECLARE_CUDA_FLOAT16_FALLBACK_UNARY(Exp, std::exp(x))
+DECLARE_CUDA_FLOAT16_FALLBACK_UNARY(Log, std::log(x))
+DECLARE_CUDA_FLOAT16_FALLBACK_UNARY(Sqrt, std::sqrt(x))
```

Comment:
I think prefixing with `DEFINE` is more correct. What do you say?

Reply:
Fixed.